src/lib.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2017 King's College London
+// created by the Software Development Team <http://soft-dev.org/>
+//
+// The Universal Permissive License (UPL), Version 1.0
+//
+// Subject to the condition set forth below, permission is hereby granted to any person obtaining a
+// copy of this software, associated documentation and/or data (collectively the "Software"), free
+// of charge and under any and all copyright rights in the Software, and any and all patent rights
+// owned or freely licensable by each licensor hereunder covering either (i) the unmodified
+// Software as contributed to or provided by such licensor, or (ii) the Larger Works (as defined
+// below), to deal in both
+//
+// (a) the Software, and
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file
+// if one is included with the Software (each a "Larger Work" to which the Software is contributed
+// by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create derivative works
+// of, display, perform, and distribute the Software and make, use, sell, offer for sale, import,
+// export, have made, and have sold the Software and the Larger Work(s), and to sublicense the
+// foregoing rights on either these or other terms.
+//
+// This license is subject to the following condition: The above copyright notice and either this
+// complete permission notice or at a minimum a reference to the UPL must be included in all copies
+// or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::marker::PhantomData;
+use std::sync::Mutex;
+
+const HOT_THRESHOLD: u16 = 10;
+
+/// What was the outcome of a `jit_control_point`?
+#[derive(Debug, PartialEq)]
+pub enum ControlOutcome {
```

Comment:
I think this enum needs a few more variants so that tests can check more aspects:
 * `StartedTracing`
 * `FinishedTracing`

Reply:
I don't want to expose that to users until we have a concrete use case for doing so. Leaking so many internal details to the outside world is dangerous IMHO.

Reply:
So why does the user need to know if a trace executed then?

I think we only return these things for testing purposes.

Reply:
I originally thought (as in the PR message at the top) that we could use the return type to remove some of the "magic" around control points. Alas, I was wrong. I now think this should just be a `void` return type, because if you think about things like black holing, it can potentially return control at very random points.

Reply:
So my next question is, how do you test this code with a void return type?

Reply:
It may make it harder: but leaking details in a public API solely for internal testing is a recipe for disaster. I think it's paramount to remember that users will come to rely on all and any details in a public API, and do things based on that which we would really prefer them not to.

Reply:
How about some kind of JIT log, which details the work of each thread, that can be parsed later?

Reply:
Maybe not parsed, it could be in-memory.

Reply:
Maybe: but that's way too much for this PR, and will require some careful thought and experimentation (including, I suspect, some performance measurements).